src/main/java/org/terasology/launcher/gui/javafx/Dialogs.java

```diff
@@ -31,17 +35,38 @@
 import java.nio.file.Path;
 import java.util.Optional;
 import java.util.concurrent.ExecutionException;
-import java.util.concurrent.FutureTask;
+import java.util.function.Supplier;
 
-public final class GuiUtils {
+public final class Dialogs {
 
-    private static final Logger logger = LoggerFactory.getLogger(GuiUtils.class);
+    private static final Logger logger = LoggerFactory.getLogger(Dialogs.class);
 
-    private GuiUtils() {
+    private Dialogs() {
+    }
+
+    private static <T> T runOnEventThread(Supplier<T> producer) {
+        T result = null;
+        if (Platform.isFxApplicationThread()) {
+            result = producer.get();
+        } else {
+            final Task<T> task = new Task<T>() {
+                @Override
+                protected T call() throws Exception {
+                    return producer.get();
+                }
+            };
+            Platform.runLater(task);
```

Comment:
and the main reason for this Platform.runLater is that we're using this with [Dialog.showAndWait](https://openjfx.io/javadoc/15/javafx.controls/javafx/scene/control/Dialog.html#showAndWait()), which says

>  This method must be called on the JavaFX Application thread. Additionally, it must either be called from an input event handler or from the run method of a Runnable passed to Platform.runLater. It must not be called during animation or layout processing.

Reply:
and while `showMessageDialog` has no return value, we're _also_ using this method to remove the duplication in `chooseDirectoryDialog`. *That* is where the `isFxApplicationThread` switch comes from, and that's why we need a Task with a return value.

Reply:
Yep, that's basically the train of thought here - including the part further above about the dialogs being modal. We had non-modal dialogs in the past, and they kept vanishing behind the launcher window and thus were often missed by users - so far that they kept working with the launcher which was trying to tell that it has crashed 🙈 

Maybe we can move to a more modern notification area where these messages show up in a non-blocking way instead. And only use modal dialogs for hard crashes from which we cannot recover.